ffi/src/lib.rs

```diff
@@ -1,173 +1,381 @@
+    fn as_ref(&'itm self) -> Self::Target;
+}
+
+/// Conversion from a type that implements [`ReprC`].
+pub trait TryFromReprC<'itm>: Sized + 'itm {
+    /// Robust C ABI compliant representation of [`Self`]
+    type Source: ReprC + Copy;
+
+    /// Type into which state can be stored during conversion. Useful for returning
+    /// non-owning types but performing some conversion which requires allocation.
+    /// Serves similar purpose as does context in a closure
+    type Store: Default;
 
-// NOTE: Using `u32` to be compatible with WebAssembly.
-// Otherwise `u8` should be sufficient
-/// Type of the handle id
-pub type HandleId = u32;
+    /// Convert from [`Self::Source`] into [`Self`].
+    /// Transferring ownership over FFI is not permitted, except for opaque pointer types
+    ///
+    /// # Errors
+    ///
+    /// * [`FfiResult::ArgIsNull`]          - given pointer is null
+    /// * [`FfiResult::UnknownHandle`]      - given id doesn't identify any known handle
+    /// * [`FfiResult::TrapRepresentation`] - given value contains trap representation
+    ///
+    /// # Safety
+    ///
+    /// All conversions from a pointer must ensure pointer validity beforehand
+    unsafe fn try_from_repr_c(
+        source: Self::Source,
+        store: &'itm mut Self::Store,
+    ) -> Result<Self, FfiResult>;
+}
+
+/// Type that can be used as an out-pointer
+pub trait OutPtr: ReprC {
+    /// Return `false` if any of the out-pointers in `Self` are null, otherwise `true`
+    fn is_valid(&self) -> bool;
+}
+
+/// Conversion into a type that can be converted to an FFI-compatible [`ReprC`] type
+/// Except for opaque pointer types, ownership transfer over FFI is not permitted
+pub trait IntoFfi: Sized {
+    /// The resulting type after conversion
+    type Target: ReprC;
+
+    /// Convert from [`Self`] into [`Self::Target`]
+    fn into_ffi(self) -> Self::Target;
+}
+
+/// Type that can be returned from an FFI function as an out pointer function argument
+pub trait FfiOutput {
+    /// Type used to represent function return value as an out pointer function argument
+    type OutPtr: OutPtr;
 
-/// FFI compatible tuple with 2 elements
-#[repr(C)]
-#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord)]
-pub struct Pair<K, V>(pub K, pub V);
+    /// Try to write [`Self`] into [`Self::OutPtr`] and return whether or not it was successful
+    ///
+    /// # Errors
+    ///
+    /// * [`FfiResult::ArgIsNull`] - if any of the out-pointers in [`Self::OutPtr`] is set to null
+    ///
+    /// # Safety
+    ///
+    /// All conversions from a pointer must ensure pointer validity beforehand
+    unsafe fn write(self, dest: Self::OutPtr) -> Result<(), FfiResult>;
+}
 
 /// Result of execution of an FFI function
 #[derive(Debug, Clone, Copy, PartialEq, Eq)]
-// NOTE: Enum is `repr(i32)` becasuse WebAssembly supports only
-// u32/i32, u64/i64 natively. Otherwise, `repr(i8)` would suffice
-#[repr(i32)]
+#[repr(i8)]
```

Comment:
Why where you able to make this change?

Reply:
I'm not, I decided against introducing wasm specifics at the moment. This will be restored

Reply:
https://github.com/hyperledger/iroha/issues/2511